Week4PalindromeLL.c

```diff
@@ -0,0 +1,221 @@
+/*
+Question: Function to check if a singly linked list is palindrome
+Solution(Optimised Approach):
+
+1. Given a linked list, find the middle element of the linked list
+  1.1 if the number of elements in linked list is odd 
+  		odd=true
+  		else
+  		odd=false
+  		
+2. Create a temporary variable Node* secondpart
+   if odd
+   		secondpart=mid->next;
+   	else
+   		secondpart=mid;
+   		
+3. Reverse the linked list with secondpart as the head
+4. Compare the two lists for palindrome test
+	head1=start
+	head2=secondpart
+	if true:
+	linked list is palindrome
+	else
+	linked list is not a palindrome
+5. Reverse the seconpart again to restore the linked list.
+
+Time Complexity: O(n)
+Space Complexity: O(1)
+*/
+#include<stdbool.h>
+#include<stdio.h>
+#include<stdlib.h>
+
+struct Node
+{
+	char ele;
+	struct Node *next;
+}*start;
+
+struct Node* createNode(char);
+bool checkPalindrome(struct Node * );
+struct Node* reverse(struct Node*);
+bool compareLists(struct Node* , struct Node *);
+
+void insert(char ele)
+{
+	if(!start)
+		start=createNode(ele);
+		
+	else
+	{
+		struct Node * temp=start;
+	
+		while(temp->next)
+			temp=temp->next;
+		temp->next=createNode(ele);	
+	}
+}
+
+struct Node* createNode(char ele)
+{
+	struct Node* newNode=(struct Node*)malloc(sizeof(struct Node));
+	newNode->ele=ele;
+	newNode->next=NULL;
+	return newNode;
+}
+
+bool checkPalindrome(struct Node * start)
+{
+	if(!start || !start->next)
+		return true;
+ 	
+	bool odd=false;
+	struct Node *fast_ptr=start;
+	struct Node *slow_ptr=start;
+	struct Node *prev_slow_ptr;
+	
+	while(fast_ptr && fast_ptr->next)
+	{   prev_slow_ptr=slow_ptr;
+		slow_ptr=slow_ptr->next;
+		fast_ptr=fast_ptr->next->next;
+	}
+
+	if(fast_ptr)
+	  odd=true;
```

Comment:
Idea to solve this problem without keeping track of whether or not it is odd
Even list
    1>2*>2>1 -> slowPrev =   2
    break into two lists (1>2) (2>1)
    swap the second list (1>2) (1>2)
    compare until the **left** list runs out
odd list
    1>2*>3>2>1 -> slowPrev = 2
    break into two lists (1>2) (3>2>1)
    swap the second list (1>2) (1>2>3)
    compare until the left list runs out

Reply:
This is a really nice strategy.
Thanks.